source/fuzz/added_function_reducer.cpp

```diff
@@ -0,0 +1,284 @@
+  // which the AddFunction transformation of interest has been simplified to use
+  // the version of the added function that appears in |reduced_binary|.
+  std::vector<uint32_t> binary_out;
+  protobufs::TransformationSequence transformation_sequence_out;
+  ReplayAdaptedTransformations(reduced_binary, &binary_out,
+                               &transformation_sequence_out);
+  return {AddedFunctionReducerResultStatus::kComplete, std::move(binary_out),
+          std::move(transformation_sequence_out), num_reduction_attempts_};
+}
+
+bool AddedFunctionReducer::InterestingnessFunctionForReducingAddedFunction(
+    const std::vector<uint32_t>& binary_under_reduction,
+    const std::unordered_set<uint32_t>& irrelevant_pointee_global_variables) {
+  uint32_t counter_for_shrinker_interestingness_function =
+      num_existing_shrink_attempts_ + num_reduction_attempts_;
+  num_reduction_attempts_++;
+
+  // The reduced version of the added function must be limited to accessing
+  // global variables appearing in |irrelevant_pointee_global_variables|.  This
+  // is to guard against the possibility of spirv-reduce changing a reference
+  // to an irrelevant global to a reference to a regular global variable, which
+  // could cause the added function to change the semantics of the original
+  // module.
+  auto ir_context =
+      BuildModule(target_env_, consumer_, binary_under_reduction.data(),
+                  binary_under_reduction.size());
+  assert(ir_context != nullptr && "The binary should be parsable.");
+  for (auto& type_or_value : ir_context->module()->types_values()) {
+    if (type_or_value.opcode() != SpvOpVariable) {
+      continue;
+    }
+    if (irrelevant_pointee_global_variables.count(type_or_value.result_id())) {
+      continue;
+    }
+    if (!ir_context->get_def_use_mgr()->WhileEachUse(
+            &type_or_value,
+            [this, &ir_context](opt::Instruction* user,
+                                uint32_t /*unused*/) -> bool {
+              auto block = ir_context->get_instr_block(user);
+              if (block != nullptr &&
+                  block->GetParent()->result_id() == GetAddedFunctionId()) {
+                return false;
+              }
+              return true;
+            })) {
+      return false;
+    }
+  }
+
+  // It must be possible to successfully apply all the transformations, with
+  // the transformation at index |index_of_add_function_transformation_|
+  // simplified to use the version of the added function from
+  // |binary_under_reduction|.
+  std::vector<uint32_t> binary_out;
+  protobufs::TransformationSequence modified_transformations;
+  ReplayAdaptedTransformations(binary_under_reduction, &binary_out,
+                               &modified_transformations);
+  if (transformation_sequence_in_.transformation_size() !=
+      modified_transformations.transformation_size()) {
+    return false;
+  }
```

Comment:
It looks like this condition is almost asserted in `ReplayAdaptedTransformations`. This is fine, just wanted to point it out in case it was an oversight. I.e. maybe we should also assert false here.

Reply:
I have added comments to clarify why we can legitimately return false here.